### PR TITLE
TMA- 712 actions now fail when unspecified param is acessed

### PR DESCRIPTION
**Pull request by Hahihula**
add params checking against the specification in all actions

params specification in actions now works

### CODE REVIEW

File: spec/unit/helpers/check_helper_spec.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  end
+  it 'verifies required' do
+    PARAMS = GoodData::LCM2::BaseAction.define_params(self) do
+      description 'Testing param one'
+      param :test_param_one, instance_of(GoodData::LCM2::Type::StringType), required: true
+    end
+    expect { GoodData::LCM2::Helpers.check_params(PARAMS, params) }.to raise_error(/Mandatory/)
+  end
+  it 'checks types' do
+    PARAMS_3 = GoodData::LCM2::BaseAction.define_params(self) do
```
Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Similar code found in check_helper_spec.rb:27, check_helper_spec.rb:35 (mass = 40)

File: spec/unit/helpers/check_helper_spec.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      description 'Testing param four'
+      param :test_param_four, instance_of(GoodData::LCM2::Type::StringType), required: false
+    end
+    expect { GoodData::LCM2::Helpers.check_params(PARAMS_3, params) }.to raise_error(/has invalid type/)
+  end
+
+  it 'fails when unspecified variable is acessed' do
+    PARAMS_4 = GoodData::LCM2::BaseAction.define_params(self) do
```
Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Similar code found in check_helper_spec.rb:27, check_helper_spec.rb:35 (mass = 40)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 3]
```diff
@@ -50,8 +111,10 @@ class << self
         def call(params)
           client = params.gdc_gd_client
           domain_name = params.organization || params.domain
+          fail "Either organisation or domain has to be specified in params" unless domain_name
           domain = client.domain(domain_name) if domain_name
           project = client.projects(params.gdc_project) || client.projects(params.gdc_project_id)
+          fail "Either project or project_id has to be specified in params" unless project
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Tests 'project' at least 4 times (RepeatedConditional)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 4]
```diff
@@ -48,7 +101,9 @@ def version
         def call(params)
           client = params.gdc_gd_client
           domain_name = params.organization || params.domain
+          fail "Either organisation or domain has to be specified in params" unless domain_name
           project = client.projects(params.gdc_project) || client.projects(params.gdc_project_id)
+          fail "Either project or project_id has to be specified in params" unless project
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Tests 'project' at least 4 times (RepeatedConditional)

File: lib/gooddata/lcm/helpers/check_helper.rb [hunk 5]
```diff
@@ -30,7 +30,7 @@ def check_params(specification, params)
               end
 
               unless type.check(value)
-                fail "Parameter '#{param_name}' has invalid type, expected: #{type}"
+                fail "Parameter '#{param_name}' has invalid type, expected: #{type}, got #{value.class}"
```
Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Calls 'value.class' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/lcm2.rb [hunk 6]
```diff
@@ -311,9 +325,11 @@ def perform(mode, params = {})
             GoodData.logger.info("Running #{action.name} action ...")
 
             # Check if all required parameters were passed
+            params.clear_filters
             BaseAction.check_params(action.const_get('PARAMS'), params)
-
+            params.setup_filters(action.const_get('PARAMS'))
```
Comment (yenkins) [inline lib/gooddata/lcm/lcm2.rb]: Calls 'action.const_get('params')' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/lcm2.rb [hunk 7]
```diff
@@ -311,9 +325,11 @@ def perform(mode, params = {})
             GoodData.logger.info("Running #{action.name} action ...")
 
             # Check if all required parameters were passed
+            params.clear_filters
```
Comment (yenkins) [inline lib/gooddata/lcm/lcm2.rb]: Calls 'params.clear_filters' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/lcm2.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/types/class/gd_logger.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
+    module Type
+      class GdLogger < ClassType
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_logger.rb]: Has no descriptive comment (IrresponsibleModule)

File: lib/gooddata/lcm/types/class/gd_logger.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_logger.rb]: Has the name 'lcm2' (UncommunicativeModuleName)

File: lib/gooddata/lcm/types/class/gd_logger.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+    module Type
+      class GdLogger < ClassType
+        CATEGORY = :class
+
+        def check(value)
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_logger.rb]: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction)

File: lib/gooddata/lcm/types/class/gd_product.rb [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
+    module Type
+      class GdProductType < ClassType
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_product.rb]: Has no descriptive comment (IrresponsibleModule)

File: lib/gooddata/lcm/types/class/gd_product.rb [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_product.rb]: Has the name 'lcm2' (UncommunicativeModuleName)

File: lib/gooddata/lcm/types/class/gd_product.rb [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+    module Type
+      class GdProductType < ClassType
+        CATEGORY = :class
+
+        def check(value)
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_product.rb]: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction)

File: lib/gooddata/lcm/types/class/gd_project.rb [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
+    module Type
+      class GdProjectType < ClassType
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_project.rb]: Has no descriptive comment (IrresponsibleModule)

File: lib/gooddata/lcm/types/class/gd_project.rb [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_project.rb]: Has the name 'lcm2' (UncommunicativeModuleName)

File: lib/gooddata/lcm/types/class/gd_project.rb [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+    module Type
+      class GdProjectType < ClassType
+        CATEGORY = :class
+
+        def check(value)
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/gd_project.rb]: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction)

File: lib/gooddata/lcm/types/class/smart_hash.rb [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
+    module Type
+      class GdSmartHashType < ClassType
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/smart_hash.rb]: Has no descriptive comment (IrresponsibleModule)

File: lib/gooddata/lcm/types/class/smart_hash.rb [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../base_type'
+
+require_relative 'class'
+
+module GoodData
+  module LCM2
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/smart_hash.rb]: Has the name 'lcm2' (UncommunicativeModuleName)

File: lib/gooddata/lcm/types/class/smart_hash.rb [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+        #   param :username, instance_of(Type::StringType), required: true
+        #
+        #   description 'Password used for connecting to GD'
+        #   param :password, instance_of(Type::StringType), required: true
+        #
+        #   description 'Hostname'
+        #   param :hostname, instance_of(Type::StringType), required: false, default: 'secure.gooddata.com'
+        # end
+
+        def check(value)
```
Comment (yenkins) [inline lib/gooddata/lcm/types/class/smart_hash.rb]: Doesn't depend on instance state (maybe move it to another class?) (UtilityFunction)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/714/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/548/ : FAILURE in 1m 07s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/825/ : SUCCESS in 35s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/715/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/549/ : SUCCESS in 40s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/826/ : SUCCESS in 36s

**Comment by kubamahnert** [comment]
extended test

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Pull request closed without merging**